### PR TITLE
Add power/exponent to math operations

### LINKED ISSUE #613: Add ** (exponentiation/power) to ? math operations
**Is your feature request related to a problem? Please describe.**
The built-in math/logical calculation/conversion command `? <expression>` is quite handy. Unfortunately, in several occasions when calculating an offset, the fun is interrupted due to the missing exponentiation/power capabilities of `?`.

**Describe the solution you'd like**
Because `^` performs XOR, `**` can be used as symbol for exponentiation/power as it is what it's used in built-in bash and zsh math capabilities for this same purpose (zsh also accepts negative exponents):
```sh
$ rz --
> ?vi 3**6
# 729
> ? 2**-1 ~fvalue[1]
# 0.5
```

**Describe alternatives you've considered**
Currently bitwise left shift could be used with 2, but no shortcuts (other than multiplying the number, or parentheses operations, with itself multiple times) exist for the other numbers/parentheses operations 💔 

**Additional context**
🤷

### PR DESCRIPTION
**Pull request by DhruvaG2000**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed) see https://github.com/rizinorg/book/pull/87

**Detailed description**
Can use ``**`` operator for finding the power of an integer

**Test plan**
To implement, 
```c
[0x00000000]> ? 2**3
int32   8
uint32  8
hex     0x8
octal   010
unit    8
segment 0000:0008
string  "\b"
fvalue  8.0
float   0.000000f
double  0.000000
binary  0b00001000
trits   0t22
```

**Closing issues**
 Closes https://github.com/rizinorg/rizin/issues/613

Signed-off-by: Dhruva Gole <goledhruva@gmail.com>

### CODE REVIEW

**Comment by DhruvaG2000** [comment]
Facing 2 difficulties which I will try to tackle:
- ~unable to use ``e`` or ``E`` or  ``**``. Not sure why only ``!`` works~ `**` is the new power Operator.
- Will not work for decimal point powers like ``0.5`` or ``1.5`` etc. Not sure how this can be done. Should I include math header and use the ``pow()`` simply?

**Comment by Rot127** [comment]
Also please add some tests in `test/db/cmd/cmd_help`

**Comment by Rot127** [comment]
> unable to use e or E or **. Not sure why only ! works

Letters get treated as `0` for me. So they are parsed somehow.
```
[0x00000000]> ? e
int32   0
uint32  0
...
```
Maybe follow the code path from the `?` command handler and check where letters get replaced or skipped.

**Comment by Rot127** [comment]
Tested with https://github.com/rizinorg/rizin/pull/2220/commits/c7102afc256eb849517e8e1aa475bb1d8ed801b8

```bash
[0x00000000]> ? 2!0
int32   2
# 2!0 != 1

[0x00000000]> ? 2!-1
# freezes. Probably an integer underflow and loops 4billion times :)

[0x00000000]> ? 2!1.4
int32   2 
# If you do not want to support floating point numbers yet, you can print a warning with `RZ_LOG_WARN()`
```

Please also add test cases once you fixed them.

**Review by wargio — CHANGES_REQUESTED** [review]
Please add tests also for negative (error, no need to implement it) and zero case (needed).

File: librz/include/rz_util/rz_num.h [hunk 1]
```diff
@@ -23,6 +23,7 @@ typedef enum {
 	RNCPLUS = '+',
 	RNCMINUS = '-',
 	RNCMUL = '*',
+	RNCEXP = '**',
```
Comment (ret2libc) [inline librz/include/rz_util/rz_num.h]: I think it is not really needed to put this here with any `=... `. In particular, this seems wrong as you can just assign one character to the enum value.

Comment (DhruvaG2000) [reply librz/include/rz_util/rz_num.h]: Yes, true. Then I am changing this to capital `E` is that okay?

File: librz/util/calc.c [hunk 2]
```diff
@@ -379,6 +397,11 @@ static RzNumCalcToken get_token(RzNum *num, RzNumCalc *nc) {
 	case '&':
 	case '|':
 	case '*':
+		if (cin_get(num, nc, &c) && c == '*') {
+			return nc->curr_tok = RNCEXP;
+		}
+		cin_putback(num, nc, c);
+		return nc->curr_tok = (RzNumCalcToken)ch;
```
Comment (ret2libc) [inline librz/util/calc.c]:
```suggestion
		return nc->curr_tok = RNCMUL;
```

File: librz/util/calc.c [hunk 3]
```diff
@@ -186,6 +186,23 @@ static RzNumCalcValue term(RzNum *num, RzNumCalc *nc, int get) {
...
+				for (int i = 1; i < (int)d.d; i++) {
+					left = Nmul(exp_left, left);
+				}
+			} else if ((int)d.d < 0) {
+				for (int i = -1; i >= (int)d.d; i--) {
+					left = Ndiv(left, exp_left);
+				}
+			} else {
+				left = Ndiv(left, exp_left);
+			}
```
Comment (ret2libc) [inline librz/util/calc.c]: The logic code should go in a `Nexp` function, similarly to how it's done for all the other operations.

File: librz/util/calc.c [hunk 4]
```diff
@@ -186,6 +186,23 @@ static RzNumCalcValue term(RzNum *num, RzNumCalc *nc, int get) {
 				return d;
 			}
 			left = Ndiv(left, d);
+		} else if (nc->curr_tok == RNCEXP) {
```
Comment (ret2libc) [inline librz/util/calc.c:208]: Whether to place this here or in `expr` or in `term` depends on which precedence we want to give to the exponential operation.

Could you put some tests with more complicated expressions so we can check it works as we'd want?

In python, the following is true:
```
>>> 10 ** 2 * 3
300
>>> 10 ** 2 / 3
33.333333333333336
>>> 10 ** 2 + 3
103
>>> 
```

File: librz/util/calc.c [hunk 5]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 	return n;
 }
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
```
Comment (ret2libc) [inline librz/util/calc.c:136]: instead of using `(int)v.d` I *think* you can just use `v.n`, which is essentially the integer representation of the same number.

File: test/db/cmd/cmd_help [hunk 6]
```diff
@@ -214,6 +214,158 @@ trits   0t2
...
+CMDS=? 10**-2
+EXPECT=<<EOF
+int32   0
+uint32  0
+hex     0x0
+octal   00
+unit    0
+segment 0000:0000
+string  "\0"
+fvalue  0.1
```
Comment (ret2libc) [inline test/db/cmd/cmd_help]: this is wrong?

File: librz/util/calc.c [hunk 7]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
...
+	RzNumCalcValue exp_n = n;
+	if (v.d - (int)v.d) {
+		RZ_LOG_WARN("floating point powers not yet supported")
```
Comment (ret2libc) [inline librz/util/calc.c]:
```suggestion
		RZ_LOG_WARN("floating point powers not yet supported\n")
```

File: librz/util/calc.c [hunk 8]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
...
+	}
+	if ((int)v.d > 0) {
+		for (int i = 1; i < (int)v.d; i++) {
+			n = Nmul(exp_n, n);
+		}
+	} else if ((int)v.d < 0) {
+		for (int i = -1; i >= (int)v.d; i--) {
+			n = Ndiv(n, exp_n);
+		}
```
Comment (ret2libc) [inline librz/util/calc.c]: this seems a bit wrong. Try to play a bit with negative exponents, I see wrong results:
```
? 10**-5
```

To compute the negative I would do `1 / Nexp(n, -exp)` or something equivalent.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/util/calc.c (hunk 7 above)
Comment (wargio) [reply librz/util/calc.c]: add also `;` (semicolumn)

**Comment by DhruvaG2000** [comment]
I have made the necessary corrections.

File: librz/util/calc.c [hunk 9]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 	return n;
 }
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
+	RzNumCalcValue exp_n = n;
+	if (v.d - (int)v.n) {
```
Comment (ret2libc) [inline librz/util/calc.c]: use `(st64)` then instead of `(int)`

File: librz/util/calc.c [hunk 10]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
 	return n;
 }
 
+static inline RzNumCalcValue Nexp(RzNumCalcValue n, RzNumCalcValue v) {
+	RzNumCalcValue exp_n = n;
```
Comment (ret2libc) [inline librz/util/calc.c:137]: any reason for having this new variable instead of just using `v`?

File: test/db/cmd/cmd_help [hunk 11]
```diff
@@ -214,6 +214,177 @@ trits   0t2
...
+CMDS=? 10**-2
+EXPECT=<<EOF
+int32   0
+uint32  0
+hex     0x0
+octal   00
+unit    0
+segment 0000:0000
+string  "\0"
+fvalue  0.0
```
Comment (ret2libc) [inline test/db/cmd/cmd_help:285]: this is wrong. Please review the code for negative exponents.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/util/calc.c (hunk 10 above)
Comment (ret2libc) [reply librz/util/calc.c:137]: Ignore this comment, I think I misread it.

File: test/db/cmd/cmd_help (hunk 11 above)
Comment (DhruvaG2000) [reply test/db/cmd/cmd_help:285]: This seems to be an issue concerning the overall code and how it handles floating point. So I am marking as resolved for now.

File: librz/util/calc.c [hunk 12]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
...
+	if (v.d - (st64)v.n) {
+		RZ_LOG_WARN("floating point powers not yet supported\n");
+	}
+	if ((st64)v.n > 0) {
+		for (st64 i = 1; i < (st64)v.n; i++) {
+			n = Nmul(exp_n, n);
+		}
+	} else if ((st64)v.n < 0) {
+		for (st64 i = 1; i > (st64)v.n; i--) {
+			n = Ndiv(n, exp_n);
```
Comment (ret2libc) [inline librz/util/calc.c:147]: I don't understand this... the first time it is executed it does `n / n`, which is ofc 1. I think it should start by doing `1 / n` and go on from there.

File: librz/util/calc.c [hunk 13]
```diff
@@ -133,6 +133,25 @@ static inline RzNumCalcValue Ndiv(RzNumCalcValue n, RzNumCalcValue v) {
...
+		}
+	} else {
+		n = Ndiv(n, exp_n);
```
Comment (ret2libc) [inline librz/util/calc.c:150]: same here. Isn't this just the number 1?

**Review by ret2libc — APPROVED** [review]

**Comment by XVilka** [comment]
```


[XX] db/esil/arm_16 cmp eq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x00000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp eq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x40000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp neq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_16 cmp neq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=16
aei
aeim
ar sp=0x00108000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x7800000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found
```

**Comment by ret2libc** [comment]
@DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

**Comment by DhruvaG2000** [comment]
> @DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

Yes I am monitoring the results, however I fail to understand how my code could be causing any of these errors.
```
[XX] db/esil/arm_32 cmp eq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp eq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar r1=0x10
ar cpsr=0x40000000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x60000000
+0x5a00000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp neq changed zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
ar cpsr=0xffffffff
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x86ffffff10000000

-- stderr
rz_config_set: variable 'esil.debug' not found



[XX] db/esil/arm_32 cmp neq unchanged zf
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e esil.debug=true
e asm.arch=arm
e asm.bits=32
aei
aeim
ar sp=0x00108000
.arf
wa cmp r1, 16
aes
?v cpsr & 0xf0000000
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0x80000000
+0x7800000000000000

-- stderr
rz_config_set: variable 'esil.debug' not found
```
What exactly is ``esil.debug``?

**Comment by DhruvaG2000** [comment]
> @DhruvaG2000 I'm re-running the CI with updated code. Please monitor the results and if there are failing tests address them, otherwise we can't merge.

Seems like everything was fine until here: https://github.com/rizinorg/rizin/pull/2220/commits/7197ce549cc2e762e67b03ac180ed9d5651b37ce 
Probably something I added after this is causing the issues. not sure if it's some syntax error in the test files or the C source code part.

[update]
The issue was in the c code ``calc.c``. Moving the ``*`` case below did the trick. Let's see (and hope nth else breaks) now.

**Review by ret2libc — APPROVED** [review]
